e2e/app/element/base-element.ts

```diff
@@ -7,69 +8,69 @@ import {ClickOptions, ElementHandle, Page, WaitForSelectorOptions} from 'puppete
 export default class BaseElement {
 
   static asBaseElement(page: Page, elem: ElementHandle): BaseElement {
-    return new BaseElement(page, elem);
+    const baseElement = new BaseElement({puppeteerPage: page});
+    baseElement.element = elem;
+    return baseElement;
   }
 
-  protected readonly page: Page;
-  protected css: string;
-  protected xpath: string;
-  protected element: ElementHandle;
+  protected _element: ElementHandle;
+  protected readonly _xpath;
+  protected readonly _page: Page;
+  protected readonly _container: Container;
 
-  constructor(aPage: Page, aElement?: ElementHandle) {
-    this.page = aPage;
-    this.element = aElement || undefined;
+  constructor(pageOptions: {puppeteerPage: Page, container?: Container},
+              selector: {xpath?: string, testId?: string} = {}) {
+    this._page = pageOptions.puppeteerPage;
+    this._container = pageOptions.container || undefined;
```

Comment:
If you have to check everything for undefined, then you're not in a very happy place. It's better to design invariants into the objects that ensure they are always in a well-defined state.

Also, I'm a bit worried that we have the same base class for a contained vs not-contained element. What about having a high-level base class that matches all elements, and an intermediate one derived from that for elements in containers? Then you don't need all these undefined members.